tests/pytorch/test_fused_rope.py

```diff
@@ -1,38 +1,164 @@
-# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 #
 # See LICENSE for license information.
 import pytest
 import torch
-from transformer_engine.pytorch.attention import RotaryPositionEmbedding, apply_rotary_pos_emb, fused_apply_rotary_pos_emb
+from typing import Callable, Dict, Tuple, Union
+from transformer_engine.pytorch.attention import (
+    RotaryPositionEmbedding,
+    apply_rotary_pos_emb,
+)
 
-def get_atol(dtype: torch.dtype) -> float:
+
+def apply_rotary_pos_emb_thd(
+    t: torch.Tensor, cu_seqlens: torch.Tensor, freqs: torch.Tensor
+) -> torch.Tensor:
+    """A baseline implementation of applying RoPE for `thd` format.
+
+    Args:
+        t (Tensor): Input tensor T is of shape [t, h, d]
+        cu_seqlens(Tensor):  Cumulative sum of sequence lengths in a batch for `t`,
+        with shape [b + 1] and dtype torch.int32.
+        freqs (Tensor): Rotary Positional embedding tensor freq is of shape [max_s, 1, 1, d]
+
+    Returns:
+        Tensor: Shape [t, h, d]. The input tensor after applying RoPE.
+    """
+    seqlens = (cu_seqlens[1:] - cu_seqlens[:-1]).tolist()
+    return torch.cat(
+        [
+            apply_rotary_pos_emb(x.unsqueeze(1), freqs[: x.size(0)])
+            for x in torch.split(t, seqlens)
+        ]
+    ).squeeze(1)
+
+
+def get_tol(dtype: torch.dtype) -> Dict:
     if dtype == torch.bfloat16:
-        return 1e-2
+        return dict(atol=1e-2, rtol=1e-2)
     elif dtype == torch.float16:
-        return 1e-3
-    return 1e-6
+        return dict(atol=1e-3, rtol=1e-3)
+    return dict(atol=1e-5, rtol=1.3e-6)
+
+
+def _overlapping_grad(output: torch.Tensor) -> torch.Tensor:
+    return output.sum() * 2
+
+
+def _non_overlapping_grad(output: torch.Tensor) -> torch.Tensor:
+    t = torch.ones_like(output)
+    return torch.sum(output * t)
```

Comment:
What is this loss function testing? Isn't it equivalent to `output.sum()`?

Reply:
`output.sum()` produces a scalar. In order to do backward calculation, it's broadcasted to a tensor. This is "overlapping" memory. In this `_non_overlapping_grad` loss function, the multiply `output * t` will receive an overlapping grad, then the fused RoPE function will receive a non-overlapping grad.

I added two different loss functions to make sure I were reading the grad tensor correctly.

Reply:
Ah, makes sense to me then. Can you add a comment to make it easier to understand in the future?